src/main/java/net/neoforged/neoforge/client/renderstate/RegisterRenderStateModifiersEvent.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.renderstate;
+
+import java.util.function.BiConsumer;
+
+import com.google.common.reflect.TypeToken;
+import net.minecraft.client.renderer.entity.EntityRenderer;
+import net.minecraft.client.renderer.entity.LivingEntityRenderer;
+import net.minecraft.client.renderer.entity.state.EntityRenderState;
+import net.minecraft.client.renderer.entity.state.LivingEntityRenderState;
+import net.minecraft.client.renderer.state.MapRenderState;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.util.context.ContextKey;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.saveddata.maps.MapDecorationType;
+import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.neoforge.client.extensions.IRenderStateExtension;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Fired for registering modifier functions for various render state objects. Useful for gathering context for
+ * custom rendering with objects that are not your own.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterRenderStateModifiersEvent extends Event implements IModBusEvent {
+    @ApiStatus.Internal
+    public RegisterRenderStateModifiersEvent() {}
+
+    /**
+     * Registers a render state modifier for {@link EntityRenderState}s which are run after all vanilla data is
+     * extracted. Can add custom data to the map using {@link EntityRenderState#setRenderData(ContextKey, Object)}.
+     * Any subclasses of the passed renderer class will also have this modifier applied.
+     * 
+     * @param baseRenderer Entity renderer class. Any subclasses will also apply this modifier.
+     * @param modifier     The function for modifying the {@link EntityRenderState} and adding custom render data.
+     * @param <E>          The type of the entity
+     * @param <S>          The specific render state type
+     */
+    public <E extends Entity, S extends EntityRenderState> void registerEntityModifier(TypeToken<? extends EntityRenderer<E, S>> baseRenderer, BiConsumer<E, S> modifier) {
+        RenderStateExtensions.registerEntity(baseRenderer.getRawType(), modifier);
```

Comment:
Isn't this a bit unsafe? After all, if this is passed with a type-token for a class with a _specific_ generic type, this will register the modifier for _all_ renderers of that raw type. That is, like, imagine someone calls this with a `BiConsumer<A, B>` and a type-token for `SomeGenericThingy<A, B>`; this will end up registering that modifier for _all_ `SomeGenericThingy`, including, say, a `SomeGenericThingy<C, D>` -- resulting in a ClassCastException at runtime when the `BiConsumer` is executed!

Reply:
Just to follow up on this, the example would be if someone did, I dunno, something like
```gradle
event.registerEntityModifier(
    new TypeToken<LivingEntityRenderer<Mob, LivingEntityRenderState, ?>>(){}
    (entity, renderState) -> {
        Mob mob = entity; // this does not require a cast, the type of `entity` is `Mob`
        // Does something that assumes `entity` is a `Mob`
    }
);
```
This code does not have any sort of unchecked casts from the end of the person writing it, and you might imagine someone could use this if they wanted to modify all renderers registered for `Mob`s! _However_, this would in fact register that modifier for _every_ `LivingEntityRenderer`, resulting in `ClassCastException`s at runtime. This is an enormous footgun -- runtime type errors are... not great. Personally, I'd say it's best to either (a) require a `Class`, so the fundamentally unsafe behavior is locked behind an unsafe cast, or (b) do validation that the `E` and `S` in question -- recovered from the TypeToken, which is its own bit of fun -- are exactly equal to the upper bounds of the class provided, to avoid this sort of issue. The second is still sub-optimal as its a runtime validation, but at least it's a runtime error that will _always_ occur instead of risking only being discovered with the wrong mod combination.

Additionally, I would say it's worth keeping around an overload with `Class` for simplicity's sake in the cases where that is enough.